Fixed version/fewer dependencies

LawrenceWarren opened the pull request:
Hey, this is the fork I made of your program. I've been using it for a little while while doing University work and it seems bug free, so I figured I'd finally make the pull request.

I looked into using Commander and opted against using it simply because single - flags can only be one character, and it seems pretty sensible that the removal flag would be `-rm` since this is pretty much standard. The good news is that I refactored my original if/else statements into something considerably more readable and extendable, and could quite easily switch to using commander if we decided to make removal `-r` instead.

In terms of (potential) work to be done:
* The README.md still needs updating.
* I was considering making it somewhat useful for nested directories (for example, you could read the comment from the parent directory to list for `../`, or read the parent directories description for the current directory)
* Some kind of git integration could be possible. If you create a `.comments` directory in the root of a git project, it could sit within the .git folder, with all of the `.comment` files being nested within here (mimicking the same tree structure as the repository) 
* Potentially setting up some kind of `.commentsignore` file, such that you can select certain extensions or directory names to not show when listing for comments.

All or none of this seems feasible to me, as the package works quite nicely as it is now. However, it looks like I will probably have all summer to kill, so any pie in the sky ideas would work.

For now, I have a whole bunch of C++ to write for an assignment due soon followed by exams, but would like to begin more work on this project come June if you see potential. Feedback is always appreciated 🔥

rumpl commented on LICENSE-MIT:
```diff
@@ -1,22 +0,0 @@
-Copyright (c) 2012 Djordje Lukic, http://rumpl.org
```
question: why remove the license?

LawrenceWarren replied:
I went on a bit of a late night directory cleaning spree, and got carried away. I'll recommit it.

rumpl commented on package.json:
```diff
@@ -1,12 +1,19 @@
-  "description": "Give folders or directories comments and view them easy.",
-  "version": "0.1.0",
+  "description": "Set and remove comments from files and directories, and view them from the command line.",
+  "version": "1.0.0",
```
Yeah!

rumpl commented on src/cli.js:
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env node
+//TODO: add a .commentsignore file for files which you would like to ignore from the --list.
```
We could creat an issue and remove this comment, I'm not a fan of TODOs, they never go away :)

LawrenceWarren replied:
I went on a bit of a "cleaning" spree late at night, and got carried away. I'll recommit it :)

rumpl commented on src/storage.js:
```diff
@@ -0,0 +1,122 @@
+ * @param {File} dir a provided directory from the file tree.
+ */
+storage.create = function (dir) {
+  fs.mkdirSync(path.join(dir, DIRECTORY), "0755"); //?Why option "0755?"
```
`0755` sets the permissions to the directory: 
* `7` means the owner can read/write/exec into this directory
* `5` means users with the same group as the directory can read/exec into it
* `5` means all other users can read/exec into the directory

rumpl commented on src/storage.js:
```diff
@@ -0,0 +1,122 @@
+var fs = require("fs"); //FileSystem
+var path = require("path"); //Paths
+var underscore = require("underscore"); //Some functional shit
+var storage = module.exports;
```
```suggestion

var storage = module.exports;
```

rumpl commented:
One thing that bothers me is that there are no more actions (like `c list`, `c set`) but flags (`c -l`) I'm not sure I'm a fan of this.

One other thing: you can now only set comments on a file/directory, but you cannot add multiple comments, it was a nice feature I think.

rumpl commented:
There's a problem when I use a flag without giving it a folder:
```
❯ node ./src/cli.js -l
Invalid flag, please try the following:
...
```

The message is misleading, it should be something like "Please specify a path". Or even better, if no path is specified then use `"."`.

rumpl commented on src/storage.js:
```diff
@@ -0,0 +1,122 @@
+
+  var commentsFile = getCommentsFile(file); //Gets the file path
+  fs.writeFileSync(commentsFile, comment); //Writes the .comment file.
+  console.log(file + ".comment was written successfully.");
```
The `storage` should be a low-level thing that doesn't interact with the user, it's the cli that should show the information/warnings/errors to the user. The storage should only manage storage.

LawrenceWarren replied:
I fixed this in a few other places throughout the storage file as well.

rumpl commented on src/storage.js:
```diff
@@ -0,0 +1,122 @@
+ * @returns An array of filenames.
+ */
+storage.loadFiles = function (dir) {
+  return underscore.filter(fs.readdirSync(dir), function (file) {
```
We can compeltely remove the `underscore` dependency from this file if we do:
```javascript
fs.readdirSync(dir).filter(file => {
  return file !== DIRECTORY;
});
```

rumpl commented on src/storage.js:
```diff
@@ -0,0 +1,122 @@
+
+var fs = require("fs"); //FileSystem
+var path = require("path"); //Paths
+var underscore = require("underscore"); //Some functional shit
```
Please remove "shit" :)

LawrenceWarren replied:
Yeah, sorry, I need to get away from amateurish... stuff.

rumpl replied:
No worries ;)

rumpl commented on src/helpers.js:
```diff
@@ -0,0 +1,79 @@
+"use strict";
+var underscore = require("underscore");
+var helpers = module.exports;
+var colors = require("colors"); //Despite looking unused, is not unused.
```
This could be `require("colors");`, no need to define a variable to take it.

Or we could use the safe one that doesn't change the `String` prototype: 
```javascript
var colors = require('colors/safe');
consle.log(colors.brightGreen('something'));
```

rumpl commented on src/cli.js:
```diff
@@ -0,0 +1,97 @@
+ */
+
+"use strict";
+var commands = require("./commands");
```
All the `require`s should be `const`

LawrenceWarren replied:
Does this also apply to exports?

rumpl replied:
What exports are we talking about?

LawrenceWarren replied:
For instance in storage, line 16: `var storage = module.exports` could become `const storage`?

rumpl commented on src/helpers.js:
```diff
@@ -0,0 +1,79 @@
+  var pad = " "; //Padding between filename and comment.
+
+  //Calls the padding a relevant amount of times, such that the length between files and their comments is SPACING wide.
+  underscore.times(maxLine - file.length + SPACING, function () {
```
JS now has a method `repeat` so this could be simply: 
```javascript
const pad = PADDING.repeat(maxLine - file.length + SPACING);
```

rumpl commented on src/helpers.js:
```diff
@@ -0,0 +1,79 @@
+ */
+helpers.printFileComments = function (files, comments) {
+  //Gets the length of the longest filename in the array - iterators through files.
+  var maxLine = underscore.max(files, function (file) {
```
This peace of code is used twice, it can be refactored into a function: 
```javascript
function maxLength(files) {
  return files.reduce((a, b) => {
    return b.length > a ? b.length : a;
  }, 0);
}
```

And with the other changes we can completely remove `underscore` as a dependency.

rumpl commented:
One last note: it would be great if you could interactively rebase your commits to have a nicer history, tell me if you need help with this.

LawrenceWarren commented:
> One thing that bothers me is that there are no more actions (like `c list`, `c set`) but flags (`c -l`) I'm not sure I'm a fan of this.

Having messed around more with the original c codebase, I agree the -- is less intuitive/more fiddly. I've renamed the `longFlag` attribute with 'action' and removed the '--'.

> One other thing: you can now only set comments on a file/directory, but you cannot add multiple comments, it was a nice feature I think.

I actually didn't realise you could previously add 2 comments. I can re-add this for now, with some thoughts about how it could work in the future (Display each comment on a new line? Delete 1 comment instead of the full '.comment' file?)

> The message is misleading, it should be something like "Please specify a path". Or even better, if no path is specified then use `"."`.

I've fixed this - you can now do `c list` and it will automatically list `./`. I've also added listing for `../` in each directory, and fixed a bug whereby if you performed `c list` on any directory which was not the current directory, it would error. It now successfully lists the files in the specified directory.

> One last note: it would be great if you could interactively rebase your commits to have a nicer history, tell me if you need help with this.

I've read a few Medium articles about people being rebase ninjas or whatever, but have never tried it myself. I'll give it a go tomorrow (it's getting late) and leave another comment if I need it.

rumpl commented on src/helpers.js:
```diff
@@ -0,0 +1,83 @@
+ */
+
+"use strict";
+const underscore = require("underscore");
```
This one is not used any more, it can also be removed from the package.json file :)

rumpl commented:
Awesome, thanks for fixing my comments, sorry there were too many. I'll play with this a little tomorrow but it's getting close to being able to be merged.

About the rebase, no stress, it's not mandatory :)

rumpl commented:
@LawrenceWarren one last thing before merging, could you add the `package-lock.json` please? Lockfiles didn't exist last time I touched this code :D

LawrenceWarren commented:
Sorry about the late reply! I had a deadline due recently, but I'm almost clear now.
I took a look and package-lock.json is already on my branch, as far as I can see, but I took the opportunity to make some changes to the README.md anyway 🔥

rumpl commented:
No worries @LawrenceWarren, thanks for the update!

Ok merging this now 🎉

LawrenceWarren commented:
Sweet! Thanks for looking through my changes and taking time to walk me through everything. I have some online exams over the next 10 or so days, but after that I have 3 months of summer time so I'll try to work on some of those additional features, or add a test suite or something. Let me know if you think of any features you want me to work on :)

P.S., I'll work on those commit messages ;)

rumpl commented:
Yo what? https://www.npmjs.com/package/c what's that I see? A version 1.0.0 published on npm?? Awesome, thank you @LawrenceWarren for the work you did!

I added two github actions:
* one that will create a relase when a tag is pushed
* the second that will release the created tag on npm

So yeah, that's awesome!

```
$ npm i -g c
$ c version
v1.0.0
```

Your code was just released on a rare one-letter package on npm ^^ congrats!
